shacl_FetchContent.cmake

```diff
@@ -0,0 +1,170 @@
+  # Option to enable force fetching via configuration
+  option(shacl_FetchContent.${name}.override_find_package "Force fetch of ${name} instead of first calling find_package" OFF)
+
+  # copy argument list as methods like list(FIND...) don't work on ${ARGV}
+  set (args ${ARGV})
+
+  # Find where FIND_PACKAGE_ARGS is present in argument list (-1 if not present)
+  list(FIND args FIND_PACKAGE_ARGS find_pkg_index)
+  list(FIND args OVERRIDE_FIND_PACKAGE override_pkg_index)
+
+  # If override_find_package (force fetching) is turned on then always fetch, never find.
+  # If force fetching then first remove FIND_PACKAGE_ARGS and all of the arguments.
+  # Then specify OVERRIDE_FIND_PACKAGE to force future calls to find_package to find the fetched version.
+  # If not force fetching then specify FIND_PACKAGE_ARGS if it wasn't present in the argument list to 
+  # first attempt finding the dependency via a call to find_paackage before trying to fetch it.
+  # FIND_PACKAGE_ARGS also ensures that if the dependency was fetched then future calls to find_package will find the fetched dependency.
+  # If OVERRIDE_FIND_PACKAGE was specified in FetchContent_Declare then FIND_PACKAGE_ARGS is not added 
+  if (${shacl_FetchContent.${name}.override_find_package})
+    # Ignore everything after FIND_PACKAGE_ARGS
+    # FIND_PACKAGE_ARGS and its associated args must come at the end of 
+    # the FetchContent_Declare function call so we can safely ignore everything after it
+    # if ${find_pkg_index} is -1 then arg_subset is equal to args
+    list(SUBLIST args 0 ${find_pkg_index} arg_subset)
+    # if OVERRIDE_FIND_PACKAGE is not present then add it
+    if (${override_pkg_index} EQUAL -1)
+      list(APPEND arg_subset OVERRIDE_FIND_PACKAGE)
+    endif()
+  else() # if not force fetching then call find_package first
+    set(arg_subset ${args})
+    # if FIND_PACKAGE_ARGS is not present and OVERRIDE_FIND_PACKAGE was not specified then add FIND_PACKAGE_ARGS
+    if (${find_pkg_index} EQUAL -1 AND ${override_pkg_index} EQUAL -1)
+      list(APPEND arg_subset FIND_PACKAGE_ARGS)
+    endif()
+  endif()
+
+  # If the dependency uses a relative path then it uses the same server as the host project.
+  # This is useful for automated testing with gitlab CI tokens or if repos are hosted on different servers
+  # e.g. if the project is hosted on github then pull all relative dependencies from github.
+
+  # Find the GIT_REPOSITORY keyword then increment index to point to associated value
+  list(FIND arg_subset GIT_REPOSITORY git_repository_index)
+  math(EXPR git_repository_index "${git_repository_index}+1")
+  list(GET arg_subset ${git_repository_index} git_repository_url)
+
+  # Update the git URL if it was a relative URL, otherwise return it unchanged
+  get_dependency_url(${git_repository_url} updated_url)
+  list(REMOVE_AT arg_subset ${git_repository_index})
+  list(INSERT arg_subset ${git_repository_index} ${updated_url})
+
+  FetchContent_Declare(${arg_subset})
+
```

Comment:
Can we add installation support for this?

Reply:
Done